File: bench/promptgreen/tasks/t30/py/hidden_tests.py
import os
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
REF = os.path.join(HERE, "reference.py")

EXPECTED = ["53000", "60100"]


def main():
    out = subprocess.run(
        [sys.executable, REF], capture_output=True, text=True, check=True
    ).stdout

    lines = out.strip().split("\n")
    assert len(lines) == 2, f"expected exactly two output lines, got {lines!r}"

    assert lines == EXPECTED, f"expected {EXPECTED!r}, got {lines!r}"

    mean_scaled = int(lines[0])
    assert 20000 <= mean_scaled <= 100000, f"mean out of data range: {mean_scaled}"

    variance_scaled = int(lines[1])
    assert variance_scaled >= 0, f"variance must be nonnegative: {variance_scaled}"

    print("t30 hidden tests: OK")


if __name__ == "__main__":
    main()

File: bench/promptgreen/tasks/t30/hidden_tests.mjs
// Hidden test for t30 (running mean/variance). Never shown to an author.
const EXPECTED = '53000\n60100\n';

export async function run(compileFn, source) {
  const r = compileFn(source);
  const ok = r.ok === true;
  const exact = ok && r.stdout === EXPECTED;

  const lines = ok ? r.stdout.trim().split('\n') : [];
  const twoLines = lines.length === 2;

  // Edge check: variance must never be negative (a broken Welford update,
  // e.g. dividing by n-1 instead of n, or reusing the pre-update delta,
  // still tends to produce a nonnegative number here, so pin the exact
  // value AND confirm the mean line alone lands within the data range
  // [2.0, 10.0] scaled by 10000, catching an accumulator that drifted).
  let meanInRange = false;
  if (twoLines) {
    const meanScaled = parseInt(lines[0], 10);
    meanInRange = Number.isFinite(meanScaled) && meanScaled >= 20000 && meanScaled <= 100000;
  }

  const green = exact && twoLines && meanInRange;
  return {
    green,
    detail: { expected: EXPECTED, got: r.stdout, ok, exact, twoLines, meanInRange },
  };
}
